qpid-jms-client/src/main/java/org/apache/qpid/jms/JmsConnectionFactory.java

```diff
@@ -368,6 +377,19 @@ protected static URI createURI(String name) {
         return null;
     }
 
+    protected Supplier<Holder<ExecutorService>> getCompletionExecutorServiceFactory() {
+        if (this.completionThreads == 0) {
+            return null;
+        }
+        synchronized (this) {
+            if (completionExecutorServiceFactory == null) {
+                QpidJMSForkJoinWorkerThreadFactory fjThreadFactory = new QpidJMSForkJoinWorkerThreadFactory("completion thread pool", true);
+                completionExecutorServiceFactory = sharedRefCnt(() -> new ForkJoinPool(completionThreads, fjThreadFactory, null, false), ThreadPoolUtils::shutdown);
```

Comment:
> I assume thats to try and avoid creating it?

Given that we cannot rely on finalization of connection factory, I cannot pre-allocate it if there are no actual "users" ie connections. And I would like it to be correctly disposed and shutdown while every connection belonging to the connction factory got closed.

Reply:
I meant was it an attempt to avoid creating the pool until the completion executor is used. Not 'avoid creating it' in terms of pre-allocating it with the factory.

This bit is only used after the point createConnection has been called, and only if the option was set. I think its reasonable to assume the pool is to be used from then given its an explicitly set option, and I dont see anyone ever setting it unless they want the behaviour.

Given that, the mechanism still all seems rather overcomplicated. This feels like a relatively simple case, an 'if there is an existing pool, then use that, otherwise create one' check coupled with the opposing cleanup. One that should be relatively infrequently used. It seems like even a simple synchronized block with a count inside could do?

I'm also weighing this vast mechanism against against a change that enables supplying a pool via the factory extension mechanism, which would probably be something ridiculous like 5-10 lines in comparison.

Reply:
> Given that, the mechanism still all seems rather overcomplicated. This feels like a relatively simple case, an 'if there is an existing pool, then use that, otherwise create one' check coupled with the opposing cleanup. One that should be relatively infrequently used. It seems like even a simple synchronized block with a count inside could do?

Not sure, there's still a problem related disposing it: 

1. the shared/common pool should be allocated once and live forever?
2. if the answer to 1 is no, how/what is going to trigger disposing it?

The mechanism I've implemented just handle this use-case using reference counting, but in order to do it, it requires a someone to be the first owner, but at least ensure correct release of resources that could cause the whole application/class-loader to leak

Reply:
No, you would still keep track with the [connection] count and shut it down if it wasnt being used, as this does, but just without a massive amount of mechanics around it to over complicate things. It would have to be retained statically of course if wanting to share across factories...like I believe your other PR already does for example.

Reply:
Got it, so you prefer a concurrent map per configuration statically referenced by each connection?
The map entries will still be `sharedRefCnt` pools, because I need to consider each connection to be able to concurrently ref/deref the shared pool (with N threads). 
What I've implemented is already using a synchronization point while allocating the shared pool, but the concurrent mechanism is needed to ensure connections to never block while releasing, nor preventing incoming ones to acquire a new/existing pool

Reply:
No. I was thinking a super simple 'get me a pool' at connection creation that bumps a count and creates a pool if needed, and an 'Im not using the pool anymore' at connection closure, that if the pool is then unused, actually shuts it down. If anotehr connection needs a pool while that is happening, it creates a new one. I'm not seeing the need for all the complexity around that which this version adds.

Reply:
> No. I was thinking a super simple 'get me a pool' at connection creation that bumps a count and creates a pool if needed

if the pool acquisition is always handled in a synchronized block, it would work, but this version I've implemented try to save blocking different connections vs the same lock if there's already an available pool.
If this is not a problem, then I can use a simpler version

Reply:
Since it wont happen at all by default, is a relatively infrequent thing, and should be pretty quick under synchronization, it seems like always synchronizing should be fine. Creating and destroying connections is already a very synchronous task with [multiple] remote trips, this doesn't seem like it would be making a significant difference at all.